File: apps/console/src/api/semanticSearch.ts
// api/semanticSearch.ts
// Typed loader for the semantic-search surface: POST /api/v0/search/semantic.
// Bounded, repository-scoped retrieval over the persisted curated
// search-document index (keyword/semantic/hybrid modes), with a language
// facet computed over the post-filter result set. Mirrors the
// relationshipsCatalog.ts client/envelope pattern: client.post +
// EshuEnvelopeError, snake_case wire records normalized to camelCase view
// types. See go/internal/query/semantic_search.go for the wire contract
// (semanticSearchRequest / semanticSearchResponse).
import type { EshuApiClient } from "./client";
import { EshuEnvelopeError } from "./envelope";

export type SemanticSearchMode = "keyword" | "semantic" | "hybrid";

// defaultSemanticSearchLimit / defaultSemanticSearchTimeoutMs are the console's
// bounded defaults for a request that does not specify them. The backend
// requires both fields (1-100 for limit, >0 for timeout_ms); these keep every
// caller of searchSemantic within the contract without repeating literals.
export const defaultSemanticSearchLimit = 20;
export const defaultSemanticSearchTimeoutMs = 8000;
export const defaultSemanticSearchMode: SemanticSearchMode = "hybrid";

export interface SemanticSearchRequest {
  readonly repoId: string;
  readonly query: string;
  readonly mode?: SemanticSearchMode;
  readonly limit?: number;
  readonly timeoutMs?: number;
  readonly serviceId?: string;
  readonly workloadId?: string;
  readonly environment?: string;
  readonly sourceKinds?: readonly string[];
  // languages filters the bounded corpus to documents carrying one of these
  // language labels. Empty/absent means no filter. An unmatched language
  // returns an empty result set rather than an error (server contract).
  readonly languages?: readonly string[];
  readonly rerank?: boolean;
}

export interface SemanticSearchDocument {
  readonly id: string;
  readonly repoId: string;
  readonly sourceKind: string;
  readonly title: string;
  readonly path: string;
  readonly contextText: string;
  readonly labels: readonly string[];
  readonly updatedAt: string;
}

export interface SemanticSearchTruthScope {
  readonly level: string;
  readonly basis: string;
}

export interface SemanticSearchFreshness {
  readonly state: string;
}

export interface SemanticSearchResult {
  readonly rank: number;
  readonly score: number;
  readonly searchMethod: string;
  readonly document: SemanticSearchDocument;
  readonly truthScope: SemanticSearchTruthScope;
  readonly freshness: SemanticSearchFreshness;
  readonly failures: readonly string[];
}

// SemanticSearchFacets carries per-facet counts computed over the
// post-filter result set. languages is always present (never undefined) on a
// successful response, matching the server's "always present" contract.
export interface SemanticSearchFacets {
  readonly languages: Readonly<Record<string, number>>;
}

export interface SemanticSearchResponse {
  readonly query: string;
  readonly repoId: string;
  readonly mode: string;
  readonly searchMode: string;
  readonly limit: number;
  readonly timeoutMs: number;
  readonly results: readonly SemanticSearchResult[];
  readonly truncated: boolean;
  readonly indexedDocumentCount: number;
  readonly corpusLimit: number;
  readonly corpusMayBeTruncated: boolean;
  readonly retrievalState: string;
  readonly facets: SemanticSearchFacets;
}

// ---------------------------------------------------------------------------
// Wire records (snake_case, as returned by the API) — kept private to this
// module; callers only see the normalized camelCase view types above.
// ---------------------------------------------------------------------------

interface ResponseRecord {
  readonly query?: string;
  readonly repo_id?: string;
  readonly mode?: string;
  readonly search_mode?: string;
  readonly limit?: number;
  readonly timeout_ms?: number;
  readonly results?: readonly ResultRecord[];
  readonly truncated?: boolean;
  readonly indexed_document_count?: number;
  readonly corpus_limit?: number;
  readonly corpus_may_be_truncated?: boolean;
  readonly retrieval_state?: string;
  readonly facets?: { readonly languages?: Record<string, number> };
}

interface ResultRecord {
  readonly rank?: number;
  readonly score?: number;
  readonly search_method?: string;
  readonly document?: DocumentRecord;
  readonly truth_scope?: { readonly level?: string; readonly basis?: string };
  readonly freshness?: { readonly state?: string };
  readonly failures?: readonly string[];
}

interface DocumentRecord {
  readonly id?: string;
  readonly repo_id?: string;
  readonly source_kind?: string;
  readonly title?: string;
  readonly path?: string;
  readonly context_text?: string;
  readonly labels?: readonly string[];
  readonly updated_at?: string;
}

// searchSemantic calls POST /api/v0/search/semantic and returns the
// normalized response. Throws EshuEnvelopeError on a server-reported error.
export async function searchSemantic(
  client: EshuApiClient,
  req: SemanticSearchRequest,
): Promise<SemanticSearchResponse> {
  const body: Record<string, unknown> = {
    repo_id: req.repoId,
    query: req.query,
    mode: req.mode ?? defaultSemanticSearchMode,
    limit: req.limit ?? defaultSemanticSearchLimit,
    timeout_ms: req.timeoutMs ?? defaultSemanticSearchTimeoutMs,
  };
  if (req.serviceId) body.service_id = req.serviceId;
  if (req.workloadId) body.workload_id = req.workloadId;
  if (req.environment) body.environment = req.environment;
  if (req.sourceKinds && req.sourceKinds.length > 0) body.source_kinds = req.sourceKinds;
  if (req.languages && req.languages.length > 0) body.languages = req.languages;
  if (req.rerank) body.rerank = true;

  const env = await client.post<ResponseRecord>("/api/v0/search/semantic", body);
  if (env.error) throw new EshuEnvelopeError(env.error);
  return normalizeResponse(env.data ?? {});
}

function normalizeResponse(data: ResponseRecord): SemanticSearchResponse {
  return {
    query: str(data.query),
    repoId: str(data.repo_id),
    mode: str(data.mode),
    searchMode: str(data.search_mode),
    limit: num(data.limit) ?? 0,
    timeoutMs: num(data.timeout_ms) ?? 0,
    results: (data.results ?? []).map(normalizeResult),
    truncated: data.truncated === true,
    indexedDocumentCount: num(data.indexed_document_count) ?? 0,
    corpusLimit: num(data.corpus_limit) ?? 0,
    corpusMayBeTruncated: data.corpus_may_be_truncated === true,
    retrievalState: str(data.retrieval_state),
    facets: { languages: normalizeLanguageFacet(data.facets?.languages) },
  };
}

function normalizeResult(record: ResultRecord): SemanticSearchResult {
  return {
    rank: num(record.rank) ?? 0,
    score: num(record.score) ?? 0,
    searchMethod: str(record.search_method),
    document: normalizeDocument(record.document ?? {}),
    truthScope: {
      level: str(record.truth_scope?.level),
      basis: str(record.truth_scope?.basis),
    },
    freshness: { state: str(record.freshness?.state) },
    failures: record.failures ?? [],
  };
}

function normalizeDocument(record: DocumentRecord): SemanticSearchDocument {
  return {
    id: str(record.id),
    repoId: str(record.repo_id),
    sourceKind: str(record.source_kind),
    title: str(record.title),
    path: str(record.path),
    contextText: str(record.context_text),
    labels: record.labels ?? [],
    updatedAt: str(record.updated_at),
  };
}

function normalizeLanguageFacet(
  raw: Record<string, number> | undefined,
): Readonly<Record<string, number>> {
  if (!raw || typeof raw !== "object") return {};
  const result: Record<string, number> = {};
  for (const [key, value] of Object.entries(raw)) {
    const count = num(value);
    if (key && count !== undefined) result[key] = count;
  }
  return result;
}

function str(value: string | undefined): string {
  return value?.trim() ?? "";
}

function num(value: number | undefined): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

File: go/internal/query/semanticsearch/semantic_search.go
// SPDX-License-Identifier: MIT
// Copyright (c) 2025-2026 eshu-hq

package semanticsearch

import (
	"context"
	"errors"
	"net/http"
	"strings"
	"time"

	"github.com/eshu-hq/eshu/go/internal/query/queryauth"
	"github.com/eshu-hq/eshu/go/internal/query/querycontract"
	"github.com/eshu-hq/eshu/go/internal/searchbench"
	"github.com/eshu-hq/eshu/go/internal/searchdocs"
	"github.com/eshu-hq/eshu/go/internal/searchretrieval"
	"github.com/eshu-hq/eshu/go/internal/telemetry"
)

// Capability is the capability-catalog identifier this family serves under.
// It gates the route by query profile and labels every error envelope and
// truth envelope the handler writes.
//
// The string lives here, with the family that implements it, and root package
// query's capability matrix reads it from here (#6060). One declaration keeps
// the profile gate, the registered support matrix, and the wire envelopes
// naming the same capability; two literals would drift into a route the matrix
// no longer describes.
const Capability = "semantic_search.curated_retrieval"

// askSearchDataClasses is the data-class set this route requires, resolved once
// at init rather than per request. PermissionDataClassesAskSearch returns a
// fresh slice on every call, so calling it inside the handler would allocate on
// the request path for a value that never changes.
var askSearchDataClasses = queryauth.PermissionDataClassesAskSearch()

// SemanticSearchHandler exposes bounded curated search-document retrieval.
type SemanticSearchHandler struct {
	Index         SemanticSearchIndexStore
	LocalHybrid   SemanticSearchHybridStore
	ScopeResolver SemanticSearchScopeResolver
	Profile       querycontract.QueryProfile
	// SearchVectorReady optionally reports the search-vector build sweep's
	// search_vector_ready watermark. When set, the response's truth freshness
	// is downgraded (pending_search_vector cause) if the sweep has never
	// published the signal, is behind its publish cadence, or the probe
	// itself failed. Nil keeps the envelope fresh (no configured signal).
	SearchVectorReady SemanticSearchVectorReadyReader
}

// SemanticSearchIndexStore searches a persisted curated search-document index
// for one repository-scoped corpus.
type SemanticSearchIndexStore interface {
	Search(context.Context, SemanticSearchIndexQuery) (SemanticSearchIndexResult, error)
}

// SemanticSearchIndexQuery is one bounded retrieval against the curated
// search-document index. ScopeID is the resolved active ingestion scope the
// index reads; RepoID stays the authorized canonical repository identity, and
// the two differ whenever a repository has been re-ingested under a new scope.
// Both are set: a store that filters on only one of them answers outside the
// caller's grant.
//
// It is exported because SemanticSearchIndexStore is: an exported interface
// whose method signature names an unexported type cannot be implemented from
// any other package, which would leave cmd/api's wiring and root package
// query's authorization tests unable to supply a store at all (#6060).
type SemanticSearchIndexQuery struct {
	Request     searchretrieval.Request
	ScopeID     string
	RepoID      string
	SourceKinds []searchdocs.SourceKind
	// Languages filters the bounded corpus to documents whose Labels contain
	// "language:<lang>" for one of the requested languages. Empty means no filter.
	Languages []string
}

// SemanticSearchIndexResult is what a store returns for one
// SemanticSearchIndexQuery: the in-scope candidates plus the corpus bounds the
// response reports so a caller can tell a genuinely empty corpus from a
// truncated one.
//
// RetrievalState may be left empty by a store; the handler fills it with the
// mode's default before the response is built, so an empty value never reaches
// the wire. Exported for the same reason as SemanticSearchIndexQuery.
type SemanticSearchIndexResult struct {
	Candidates           []searchretrieval.Candidate
	IndexedDocumentCount int
	CorpusLimit          int
	CorpusMayBeTruncated bool
	RetrievalState       string
}

type semanticSearchRequest struct {
	RepoID      string   `json:"repo_id"`
	Query       string   `json:"query"`
	Mode        string   `json:"mode"`
	Limit       int      `json:"limit"`
	TimeoutMS   int      `json:"timeout_ms"`
	ServiceID   string   `json:"service_id,omitempty"`
	WorkloadID  string   `json:"workload_id,omitempty"`
	Environment string   `json:"environment,omitempty"`
	SourceKinds []string `json:"source_kinds,omitempty"`
	// Languages filters the corpus to documents whose Labels contain
	// "language:<lang>" for one of the requested languages. An empty slice
	// means no language filter. Any non-empty lowercased token is accepted;
	// an unmatched language returns an empty result set rather than an error.
	// The index is the source of truth for which language values exist.
	Languages []string `json:"languages,omitempty"`
	// Rerank opts the request into graph-neighborhood reranking over the
	// retrieved in-scope results. Off by default; when on, the response reports
	// the reranking state, per-result ranking basis, and recommended next calls.
	Rerank bool `json:"rerank,omitempty"`
}

// semanticSearchFacets carries per-facet counts over the already-bounded
// in-scope candidate corpus.
type semanticSearchFacets struct {
	// Languages maps each "language:<x>" label value (the "<x>" part) to the
	// count of results carrying that language in the post-filter result set.
	Languages map[string]int `json:"languages"`
}

type semanticSearchResponse struct {
	Query                    string                 `json:"query"`
	RepoID                   string                 `json:"repo_id"`
	Anchor                   searchretrieval.Anchor `json:"anchor"`
	Mode                     searchbench.Mode       `json:"mode"`
	SearchMode               string                 `json:"search_mode"`
	Limit                    int                    `json:"limit"`
	TimeoutMS                int                    `json:"timeout_ms"`
	Results                  []semanticSearchResult `json:"results"`
	Truncated                bool                   `json:"truncated"`
	FalseCanonicalClaimCount int                    `json:"false_canonical_claim_count"`
	IndexedDocumentCount     int                    `json:"indexed_document_count"`
	CorpusLimit              int                    `json:"corpus_limit"`
	CorpusMayBeTruncated     bool                   `json:"corpus_may_be_truncated"`
	RetrievalState           string                 `json:"retrieval_state"`
	// Facets carries per-dimension aggregate counts computed over the
	// post-filter result set. The block is always present (never omitted) so
	// callers can rely on the shape unconditionally.
	Facets               semanticSearchFacets  `json:"facets"`
	Rerank               *semanticSearchRerank `json:"rerank,omitempty"`
	RecommendedNextCalls []semanticSearchCall  `json:"recommended_next_calls,omitempty"`
}

type semanticSearchResult struct {
	Rank         int                         `json:"rank"`
	Score        float64                     `json:"score"`
	SearchMethod string                      `json:"search_method"`
	Document     semanticSearchDocument      `json:"document"`
	GraphHandles []semanticSearchGraphHandle `json:"graph_handles"`
	TruthScope   semanticSearchTruthScope    `json:"truth_scope"`
	Freshness    semanticSearchFreshness     `json:"freshness"`
	Failures     []searchbench.FailureClass  `json:"failures,omitempty"`
	Metadata     map[string]string           `json:"metadata,omitempty"`
	RankingBasis *semanticSearchRankingBasis `json:"ranking_basis,omitempty"`
}

type semanticSearchDocument struct {
	ID           string                      `json:"id"`
	RepoID       string                      `json:"repo_id"`
	SourceKind   searchdocs.SourceKind       `json:"source_kind"`
	Title        string                      `json:"title"`
	Path         string                      `json:"path,omitempty"`
	ContextText  string                      `json:"context_text,omitempty"`
	EntityRefs   []semanticSearchEntityRef   `json:"entity_refs,omitempty"`
	GraphHandles []semanticSearchGraphHandle `json:"graph_handles"`
	Labels       []string                    `json:"labels,omitempty"`
	UpdatedAt    time.Time                   `json:"updated_at,omitempty"`
	TruthScope   semanticSearchTruthScope    `json:"truth_scope"`
	Freshness    semanticSearchFreshness     `json:"freshness"`
	AccessScope  semanticSearchAccessScope   `json:"access_scope"`
	Provenance   semanticSearchProvenance    `json:"provenance"`
}

type semanticSearchEntityRef struct {
	ID        string `json:"id"`
	Type      string `json:"type,omitempty"`
	Name      string `json:"name,omitempty"`
	Path      string `json:"path,omitempty"`
	StartLine int    `json:"start_line,omitempty"`
	EndLine   int    `json:"end_line,omitempty"`
}

type semanticSearchGraphHandle struct {
	Kind string `json:"kind"`
	ID   string `json:"id"`
}

type semanticSearchTruthScope struct {
	Level searchdocs.TruthLevel `json:"level"`
	Basis searchdocs.TruthBasis `json:"basis"`
}

type semanticSearchFreshness struct {
	State searchdocs.FreshnessState `json:"state"`
}

type semanticSearchAccessScope struct {
	RepoID string `json:"repo_id,omitempty"`
}

type semanticSearchProvenance struct {
	SourceTable string   `json:"source_table,omitempty"`
	SourceIDs   []string `json:"source_ids,omitempty"`
}

// Mount registers semantic-search routes.
func (h *SemanticSearchHandler) Mount(mux *http.ServeMux) {
	mux.HandleFunc("POST /api/v0/search/semantic", h.search)
}

func (h *SemanticSearchHandler) profile() querycontract.QueryProfile {
	if h == nil || h.Profile == "" {
		return querycontract.ProfileProduction
	}
	return h.Profile
}

func (h *SemanticSearchHandler) search(w http.ResponseWriter, r *http.Request) {
	r, span := startQueryHandlerSpan(
		r,
		telemetry.SpanQuerySemanticSearch,
		"POST /api/v0/search/semantic",
		Capability,
	)
	defer span.End()

	if !queryauth.AllowsPermissionFeature(r.Context(), queryauth.PermissionFeatureAskSearch) {
		writeSemanticSearchError(
			w,
			r,
			http.StatusForbidden,
			querycontract.ErrorCodePermissionDenied,
			"permission denied",
		)
		return
	}
	if !queryauth.AllowsPermissionDataClasses(r.Context(), askSearchDataClasses...) {
		writeSemanticSearchError(
			w,
			r,
			http.StatusForbidden,
			querycontract.ErrorCodePermissionDenied,
			"permission denied",
		)
		return
	}

	if querycontract.CapabilityUnsupported(h.profile(), Capability) {
		querycontract.WriteContractError(
			w,
			r,
			http.StatusNotImplemented,
			"semantic search requires curated search-document retrieval",
			querycontract.ErrorCodeUnsupportedCapability,
			Capability,
			h.profile(),
			querycontract.RequiredProfile(Capability),
		)
		return
	}

	var body semanticSearchRequest
	if err := querycontract.ReadJSON(r, &body); err != nil {
		writeSemanticSearchError(w, r, http.StatusBadRequest, querycontract.ErrorCodeInvalidArgument, err.Error())
		return
	}
	body = normalizeSemanticSearchRequest(body)
	sourceKinds, err := semanticSearchSourceKinds(body.SourceKinds)
	if err != nil {
		writeSemanticSearchError(w, r, http.StatusBadRequest, querycontract.ErrorCodeInvalidArgument, err.Error())
		return
	}
	languages := semanticSearchLanguages(body.Languages)
	req, err := semanticSearchRetrievalRequest(body)
	if err != nil {
		writeSemanticSearchError(w, r, http.StatusBadRequest, querycontract.ErrorCodeInvalidArgument, err.Error())
		return
	}

	access := querycontract.RepositoryAccessFilterFromContext(r.Context())
	if access.Empty() {
		querycontract.WriteSuccess(w, r, http.StatusOK, emptySemanticSearchResponse(req), h.truthWithSearchVectorFreshness(r, req.Mode))
		return
	}
	directScopeGrant := access.AllowsDirectScopeID(body.RepoID)
	canonicalRepositoryGrant := access.AllowsCanonicalRepositoryID(body.RepoID)
	if !directScopeGrant && !canonicalRepositoryGrant {
		writeSemanticSearchError(w, r, http.StatusNotFound, querycontract.ErrorCodeNotFound, "repository not found")
		return
	}
	resolution, err := h.resolveScope(
		r.Context(),
		body.RepoID,
		access,
		directScopeGrant,
		canonicalRepositoryGrant,
	)
	if err != nil {
		if resolution.ambiguous {
			writeSemanticSearchError(w, r, http.StatusConflict, querycontract.ErrorCodeAmbiguous, err.Error())
			return
		}
		writeSemanticSearchError(w, r, http.StatusServiceUnavailable, querycontract.ErrorCodeBackendUnavailable, err.Error())
		return
	}
	if resolution.scopeID == "" {
		querycontract.WriteSuccess(w, r, http.StatusOK, emptySemanticSearchResponse(req), h.truthWithSearchVectorFreshness(r, req.Mode))
		return
	}
	req = semanticSearchCanonicalAnchorRequest(span, req, resolution.repositoryID)
	var indexResult SemanticSearchIndexResult
	backend, err := h.semanticSearchBackend(
		req,
		resolution.scopeID,
		resolution.repositoryID,
		sourceKinds,
		languages,
		&indexResult,
	)
	if err != nil {
		writeSemanticSearchError(
			w,
			r,
			http.StatusServiceUnavailable,
			querycontract.ErrorCodeBackendUnavailable,
			err.Error(),
		)
		return
	}

	retrieval, err := (searchretrieval.Runner{
		Backend: backend,
	}).Retrieve(r.Context(), req)
	if err != nil {
		annotateSemanticSearchDegradedError(r.Context(), span, err)
		status, code := semanticSearchRetrievalError(err)
		writeSemanticSearchError(w, r, status, code, err.Error())
		return
	}

	annotateSemanticSearchDegraded(r.Context(), span, req.Mode, indexResult.RetrievalState)

	querycontract.WriteSuccess(
		w,
		r,
		http.StatusOK,
		semanticSearchResponseFromRetrieval(req, retrieval, indexResult, body.Rerank),
		h.truthWithSearchVectorFreshness(r, req.Mode),
	)
}

func (h *SemanticSearchHandler) semanticSearchBackend(
	req searchretrieval.Request,
	scopeID string,
	repoID string,
	sourceKinds []searchdocs.SourceKind,
	languages []string,
	indexResult *SemanticSearchIndexResult,
) (searchretrieval.Backend, error) {
	query := SemanticSearchIndexQuery{
		Request: req,
		// The public slice is repository-bounded. ScopeID is the resolved active
		// ingestion scope; RepoID remains the authorized canonical identity.
		ScopeID:     scopeID,
		RepoID:      repoID,
		SourceKinds: sourceKinds,
		Languages:   languages,
	}
	if h.LocalHybrid != nil && (req.Mode == searchbench.ModeSemantic || req.Mode == searchbench.ModeHybrid) {
		return semanticSearchIndexBackend{
			Index:         h.LocalHybrid,
			Query:         query,
			Result:        indexResult,
			AllowSemantic: true,
		}, nil
	}
	if h.Index == nil {
		return nil, errors.New("semantic search requires the persisted search index")
	}
	return semanticSearchIndexBackend{
		Index:  h.Index,
		Query:  query,
		Result: indexResult,
	}, nil
}

type semanticSearchIndexBackend struct {
	Index         SemanticSearchIndexStore
	Query         SemanticSearchIndexQuery
	Result        *SemanticSearchIndexResult
	AllowSemantic bool
}

func (backend semanticSearchIndexBackend) Search(
	ctx context.Context,
	req searchretrieval.Request,
) ([]searchretrieval.Candidate, error) {
	if req.Mode == searchbench.ModeSemantic && !backend.AllowSemantic {
		return nil, errors.New("semantic mode requires an embedder")
	}
	query := backend.Query
	query.Request = req
	result, err := backend.Index.Search(ctx, query)
	if err != nil {
		return nil, err
	}
	if result.RetrievalState == "" {
		result.RetrievalState = defaultSemanticSearchRetrievalState(req.Mode)
	}
	if backend.Result != nil {
		*backend.Result = result
	}
	return result.Candidates, nil
}

func semanticSearchRetrievalError(err error) (int, querycontract.ErrorCode) {
	message := err.Error()
	if strings.Contains(message, "semantic mode requires an embedder") {
		return http.StatusServiceUnavailable, querycontract.ErrorCodeBackendUnavailable
	}
	if strings.Contains(message, "search retrieval backend") {
		return http.StatusInternalServerError, querycontract.ErrorCodeInternalError
	}
	return http.StatusBadRequest, querycontract.ErrorCodeInvalidArgument
}

func (h *SemanticSearchHandler) truth() *querycontract.TruthEnvelope {
	return querycontract.BuildTruthEnvelope(
		h.profile(),
		Capability,
		querycontract.TruthBasisHybrid,
		"resolved from a persisted curated search-document index",
	)
}

// truthWithSearchVectorFreshness builds the response truth envelope and, when
// a search-vector-ready reader is configured AND the resolved mode is
// vector-backed (semantic or hybrid — the same gate semanticSearchBackend
// uses to decide whether LocalHybrid's vector path is even wired in),
// downgrades it from the search-vector build sweep's search_vector_ready
// watermark so an outstanding build is attributable (pending_search_vector)
// instead of served as silently fresh. mode:"keyword" is served entirely by
// the deterministic lexical index and is never degraded by vector/index
// readiness (see semanticSearchDegradation), so it must never be downgraded
// by a pending search-vector build. Mirrors applyWinnersFreshness's call-site
// shape in supply_chain_impact_findings_handler.go: a probe failure reports
// the envelope unavailable rather than dropping the already-served results.
func (h *SemanticSearchHandler) truthWithSearchVectorFreshness(r *http.Request, mode searchbench.Mode) *querycontract.TruthEnvelope {
	truth := h.truth()
	if h.SearchVectorReady == nil || !searchVectorBackedMode(mode) {
		return truth
	}
	watermark, err := h.SearchVectorReady.SearchVectorReadyWatermark(r.Context())
	applySearchVectorFreshness(truth, watermark, err, time.Now())
	return truth
}

// searchVectorBackedMode reports whether mode retrieves through the
// search-vector index (semantic or hybrid), matching the gate
// semanticSearchBackend uses to decide whether LocalHybrid's vector path is
// engaged. mode:"keyword" is served entirely by the deterministic lexical
// index and never touches search-vector state.
func searchVectorBackedMode(mode searchbench.Mode) bool {
	return mode == searchbench.ModeSemantic || mode == searchbench.ModeHybrid
}

func writeSemanticSearchError(w http.ResponseWriter, r *http.Request, status int, code querycontract.ErrorCode, message string) {
	if querycontract.AcceptsEnvelope(r) {
		querycontract.WriteJSON(w, status, querycontract.ResponseEnvelope{Error: &querycontract.ErrorEnvelope{
			Code:       code,
			Message:    message,
			Capability: Capability,
		}})
		return
	}
	querycontract.WriteJSON(w, status, map[string]any{
		"error_code": code,
		"message":    message,
		"capability": Capability,
	})
}
